### PR TITLE
fix float_label too fast in some cases

### PR DESCRIPTION
**Pull request by gfgtdf**
fixes https://github.com/wesnoth/wesnoth/issues/4347

the fix kinda works but for some reason it created errors, when undrawing out-of-map  parts. Also i feel dirty after touching the drawing  code.

### CODE REVIEW

File: src/floating_label.hpp [hunk 1]
```diff
@@ -69,21 +66,28 @@ class floating_label
...
 
 	LABEL_SCROLL_MODE scroll() const { return scroll_; }
 
 private:
 
+	int get_time_alive(int current_time) const { return current_time - time_start_; }
 	int xpos(std::size_t width) const;
+	SDL_Point get_loc(int time);
+	surface get_surface(int time);
+	bool is_expired(bool time);
```
Comment (stevecotton) [inline src/floating_label.hpp]: `is_expired` is unused

File: src/floating_label.cpp [hunk 2]
```diff
@@ -177,29 +180,47 @@ void floating_label::draw(surface screen)
...
 	}
 
-	SDL_Rect rect = sdl::create_rect(xpos(surf_->w), ypos_, surf_->w, surf_->h);
+	SDL_Point pos = get_loc(time);
+	buf_pos_ = sdl::create_rect(pos.x, pos.y, surf_->w, surf_->h);
 	const clip_rect_setter clip_setter(screen, &clip_rect_);
-	sdl_copy_portion(screen,&rect,buf_,nullptr);
-	sdl_blit(surf_,nullptr,screen,&rect);
+	sdl_copy_portion(screen, &buf_pos_, buf_, nullptr);
+	sdl_blit(get_surface(time), nullptr, screen, &buf_pos_);
```
Comment (stevecotton) [inline src/floating_label.cpp]: Inconsistent checking for `surf_ != nullptr` - it's not checked here, but inside `get_surface` there's a check which will make that function return nullptr.

**Comment by gfgtdf** [comment]
these requests are true, but i probalby won't implement them until i found the casue of the actual error, which is that it kinda still looks worse than before. I was kinda hoping that someone woudl have an idea on that one.

**Comment by stevecotton** [comment]
If you're going to be doing a lot more on this, would it be a sensible refactor to drop the whole "undraw" logic first?

IIUC the undrawing is an optimisation from the time that few things were moving on screen - now that Wesnoth has lots of animated terrain and units doing idle animation, it might make more sense to just redraw the whole screen each frame. All of those can be turned off in the display settings, but so can the floating text.

**Comment by gfgtdf** [comment]
> If you're going to be doing a lot more on this

i really hope i don't.

> it might make more sense to just redraw the whole screen each frame

yes i really don't intend to do such huge changes right now, there is a reason why people are working on godot port rn, and the main reason for that is afaik noone really likes our current drawing engine anymore.

File: src/floating_label.cpp (hunk 2 above)
Comment (gfgtdf) [reply src/floating_label.cpp]: but a few lines above this function checks for for surf_ == nullptr

**Comment by gfgtdf** [comment]
i found the cause if the problems, appearently sdl_clit modifes the dst_rect parameter, this is ready to merge now

**Comment by gfgtdf** [comment]
also fixed the the issue mentioned above

**Pull request merged**